crates/ruff_python_formatter/src/comments/format.rs

```diff
@@ -453,41 +466,78 @@ impl Format<PyFormatContext<'_>> for FormatNormalizedComment<'_> {
 fn normalize_comment<'a>(
     comment: &'a SourceComment,
     source: SourceCode<'a>,
-) -> FormatResult<Cow<'a, str>> {
+) -> FormatResult<NormalizedComment> {
     let slice = comment.slice();
     let comment_text = slice.text(source);
 
-    let trimmed = comment_text.trim_end();
+    // Strip trailing whitespace.
+    let trimmed_text = comment_text.trim_end();
 
-    let content = strip_comment_prefix(trimmed)?;
+    let mut cursor = Cursor::new(trimmed_text);
 
-    if content.is_empty() {
-        return Ok(Cow::Borrowed("#"));
+    // Strip the leading `#`.
+    if !cursor.eat_char('#') {
+        return Err(FormatError::syntax_error(
+            "Didn't find expected comment token `#`",
+        ));
+    };
+
+    // If the comment is empty, just return the leading `#`.
+    if cursor.is_eof() {
+        return Ok(NormalizedComment::Verbatim(TextRange::at(
+            slice.start(),
+            '#'.text_len(),
+        )));
     }
 
     // Fast path for correctly formatted comments: if the comment starts with a space, or any
     // of the allowed characters, then it's included verbatim (apart for trimming any trailing
     // whitespace).
-    if content.starts_with([' ', '!', ':', '#', '\'']) {
-        return Ok(Cow::Borrowed(trimmed));
+    if matches!(cursor.first(), ' ' | '!' | ':' | '#' | '\'') {
+        return Ok(NormalizedComment::Verbatim(TextRange::at(
+            slice.start(),
+            '#'.text_len() + cursor.text_len(),
```

Comment:
You could use `cursor.start_token()` and then `cursor.token_len` to get the length instead of doing the math manually (same in other places)